lib/api/core/plugins/manifest.js

```diff
@@ -0,0 +1,108 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2018 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const
+  _ = require('lodash'),
+  { PluginImplementationError } = require('kuzzle-common-objects').errors,
+  path = require('path'),
+  semver = require('semver');
+
+/* eslint-disable no-console */
+
+/**
+ * This class initializes itself using a manifest.json file
+ *
+ * @deprecated  - the "fallback" argument is deprecated: there will be no
+ * fallback if a manifest.json file is missing in the next major release of
+ * Kuzzle. For now, the fallback is a function receiving this instance as
+ * an argument, and is in charge of updating it accordingly
+ *
+ * @param {Kuzzle} kuzzle       - Kuzzle instance
+ * @param {string} pluginPath   - Absolute path to the plugin directory
+ * @param {Function} [fallback] - Fallback function, if no manifest.json is found
+ */
+class Manifest {
+  constructor(kuzzle, pluginPath, fallback = null) {
+    this.path = pluginPath;
+    this.name = null;
+    this.privileged = false;
+
+    // Only Kuzzle 1.x should support plugins without a kuzzleVersion property set
+    this.kuzzleVersion = '>=1.0.0 <2.0.0';
+
+    let raw;
+    try {
+      raw = require(path.resolve(this.path, 'manifest.json'));
+    } catch (e) {
+      // @deprecated
+      if (fallback !== null) {
+        console.warn(`[${this.path}] Plugins without a manifest.json file are deprecated.`);
+        fallback(this);
+        return this;
+      }
+
+      throw new PluginImplementationError(e);
+    }
+
+    if (!_.isNil(raw.privileged)) {
+      if (typeof raw.privileged !== 'boolean') {
+        throw new PluginImplementationError(`[${this.path}/manifest.json] Invalid "privileged" property: expected a boolean, got a ${typeof raw.privileged}`);
+      }
+      this.privileged = raw.privileged;
+    }
+
+    if (!_.isNil(raw.kuzzleVersion)) {
+      if (typeof raw.kuzzleVersion !== 'string' || raw.kuzzleVersion.length === 0) {
+        throw new PluginImplementationError(`[${this.path}/manifest.json] Invalid "kuzzleVersion" property: expected a non-empty string`);
+      }
+      this.kuzzleVersion = raw.kuzzleVersion;
+    } else {
+      console.warn(`[${pluginPath}/manifest.json] No "kuzzleVersion" property found: assuming the target is Kuzzle v1`);
```

Comment:
Do we really need so many type checking (and thus lodash) to validate a string that will eventually be sent to compareVersion?

Reply:
There are 2 type checks:

* We need to check against null/undefined in order to display a warning and use a default semver version range, preventing the introduction of a breaking change. I indeed use lodash for that because I find `_.isNul(raw.kuzzleVersion)` easier to read than `raw.kuzzleVersion !== undefined && raw.kuzzleVersion !== null`, at no extra cost
* I then check against other types because semver is inconsistent when a non-string value is provided: it may return `false`, thus triggering a (rather imprecise) `version mismatch` error from our side, or it may throw an `invalid version` error. Handling these two cases seemed simpler to me with a simple test before using semver

Reply:
semver inconsistencies showcases:

```
false
> semver.satisfies(false, '1.2.3')
false
> semver.satisfies(true, '1.2.3')
TypeError: Invalid Version: true
    at new SemVer (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:291:11)
    at compare (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:578:10)
    at eq (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:617:10)
    at cmp (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:649:41)
    at Comparator.test (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:716:10)
    at testSet (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:1123:17)
    at Range.test (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:1115:9)
    at Function.satisfies (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:1161:16)
> semver.satisfies(12, '1.2.3')
TypeError: Invalid Version: 12
    at new SemVer (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:291:11)
    at compare (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:578:10)
    at eq (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:617:10)
    at cmp (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:649:41)
    at Comparator.test (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:716:10)
    at testSet (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:1123:17)
    at Range.test (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:1115:9)
    at Function.satisfies (/home/scottinet/git/kuzzle/node_modules/semver/semver.js:1161:16)
```

Reply:
Hm. Dang.

```
semver.satisfies('foosdflkj fsjfd jg1.2', '1.2.3')
TypeError: Invalid Version: foosdflkj fsjfd jg1.2
```

I will have to rethink that one a bit.